### PR TITLE
Fp16 nchw for cudnn-fp16 backend (support GTX 16xx GPUs)

### PR DESCRIPTION
**Pull request by ankan-ban**
For cudnn-fp16 backend: For supporting GPUs without tensor cores (e.g: GP100 and GTX 16xx series).
TODO: 
1. Figure out a way to check for tensor cores (and select nhwc vs nchw based on that).
 - done (Unfortunately involves string matching the device name. Hopefully will be fixed in some later cuda release.)
2. actually run on a gtx 16xx card and check performance.
 - done (see below). Slightly less than 2x speedup over fp32.
3. misc cleanup and clang format.
 - done
4. (optional): maybe write a fused kernel for SE layer.
 - maybe later (or maybe not needed. With current implementation SE raw nps is only ~6% lower than non-SE net, tested with 256x20 networks).

### CODE REVIEW

**Comment by ankan-ban** [comment]
Some benchmarks on GTX 1650

```
1. FP32
Benchmark final time 9.77541s calculating 2303.53 nodes per second.

2. fp16 with nhwc (current default)
Benchmark final time 9.35774s calculating 224.627 nodes per second.

3. fp16 with nchw layout - with 'TENSOR_OP_MATH' setting enabled.
Benchmark final time 8.7635s calculating 535.517 nodes per second.

4. fp16 with nchw layout - without 'TENSOR_OP_MATH' setting enabled.
Benchmark final time 8.67777s calculating 4238.88 nodes per second.
```

File: src/neural/cuda/common_kernels.cu [hunk 1]
```diff
@@ -281,10 +282,54 @@ void expandPlanes_Fp16_NHWC(half* output, const uint64_t* masks,
...
 
-__global__ void globalScale_kernel(float* output, const float* input,
-                                   const float* scaleBias,
-                                   const float* prevLayerBias, int inputSize,
-                                   int C) {
+__global__ void expandPlanes_kernel_Fp16_NCHW(half* output,
+                                              const uint64_t* masks,
+                                              const float* values, int n) {
+  // block size of 256, same mask/val for 64 consecutive threads
+  constexpr int kNumShmemElments = 256 / 64;
```
[the remark this answers is not in the document]
Comment (ankan-ban) [reply src/neural/cuda/common_kernels.cu]: Me too. likely a typo. Will fix.

File: src/neural/cuda/network_cudnn.cc [hunk 2]
```diff
@@ -184,14 +214,39 @@ class CudnnNetwork : public Network {
...
       }
+
+      // Override if forced from backend option (default: -1 means auto-select).
+      int force_nhwc = options.GetOrDefault<int>("force_nhwc", -1);
+      if (force_nhwc == 0) {
+        nhwc_ = false;
+      } else if (force_nhwc == 1) {
+        nhwc_ = true;
+      }
+
```
[the remark this answers is not in the document]
Comment (ankan-ban) [reply src/neural/cuda/network_cudnn.cc]: Thanks. Seems to work as expected.

File: lc0@exe/lc0@exe.log [hunk 3]
```diff
@@ -0,0 +1,2 @@
+﻿  layers.cc
+  lc0@exe.vcxproj -> C:\Ankan\git\ankan\lc0\build\.\lc0.exe
```
Comment (mooskagh) [inline lc0@exe/lc0@exe.log]: What is this file? :)

Comment (ankan-ban) [reply lc0@exe/lc0@exe.log]: Sorry. Likely some intermediate build file that accidentally got submitted. Will remove it.

**Comment by rajb245** [comment]
Does the merged work apply only to the cards using the GP100, i.e., the Quadro GP100 and the Tesla P100? Can similar techniques apply to the other Pascal chips, in particular, GP102 chips like the Titan X (Pascal) and Titan Xp have? NVIDIA advertises some level of fp16 acceleration on those, but I don't know enough of the implementation to know the differences.

If there's a path to accelerate performance on GP102 using similar techniques, please let me know and I'll open a feature request issue.

**Comment by ankan-ban** [comment]
Unfortunately no, other Pascal chips (gp102/gp104/gp106, etc) don't have support for fp16 math. They do support higher throughput int8 math but right now we don't have support in lc0 for int8 precision.

**Comment by jjoshua2** [comment]
5.3 support for fp16 is still missing for jetson. I tested just adding it
and it works fine to double the speeds it seems.

**Comment by borg323** [comment]
Do you know whether using fp16 with CC 6.2 (jetson TX2) is also a performance gain?

**Pull request merged**